FancyBatching.md

```diff
@@ -0,0 +1,505 @@
+    the JSON payload and not the number of documents.  We'll keep this private
+    for the time being until we can provide reasonable tuning.
+
+    - `private int RetryCount { get; set; } = 3;`  The number of manual retries
+    we'll perform for failed documents.  We default to the same as Azure.Core
+    which is 3 retries and keep it private for now while we think about how to
+    expose the tuning of these parameters.
+
+    - `private ResponseClassifier RetryClassifier { get; set; }`  This is the
+    type we use in Azure.Core to customize Retry behavior.  It will basically
+    retry on 422, 409, 503 status codes and the standard IOException and
+    RequestFailedException with status code 0.
+
+    - TODO: If we end up using PID Controllers for control stabilization, should
+    we expose their gain configuration?
+
+## Using SearchIndexingBufferedSender
+
+- `public class SearchIndexingBufferedSender<T>`
+    - Add Indexing Actions
+        - `void DeleteDocuments(IEnumerable<T>, CancellationToken)`
+        - `Task DeleteDocumentsAsync(IEnumerable<T>, CancellationToken)`
+        - `void IndexDocuments(IndexDocumentsBatch<T>, CancellationToken)`
+        - `Task IndexDocumentsAsync(IndexDocumentsBatch<T>, CancellationToken)`
+
+            Note these are using a batch which we already have conveniences for
+            building up.
+
+        - `void MergeDocuments(IEnumerable<T>, CancellationToken)`
+        - `Task MergeDocumentsAsync(IEnumerable<T>, CancellationToken)`
+        - `void MergeOrUploadDocuments(IEnumerable, CancellationToken)`
+        - `Task MergeOrUploadDocumentsAsync(IEnumerable<T>, CancellationToken)`
+        - `void UploadDocuments(IEnumerable<T>, CancellationToken)`
+        - `Task UploadDocumentsAsync(IEnumerable<T>, CancellationToken)`
+
+        We're not returning any data from these operations.  They may submit
+        something immediately or we might be waiting for a flush.
+
+        The provided cancellation token will be used for any requests sent, any
+        notifications raised, etc.  If a flush happens when the timer elapses,
+        then we'll use the `AutoFlushIntervalCancellationToken` instead.  A 
+        customer could reuse a token or link them together for fine grained
+        control.
+
+        We're also not including the `IndexDocumentsOptions` parameter because
+        that currently only serves to configure exception behavior which we're
+        doing exclusively through events.
+
+        - Delete by key
+
+            Add customization for "delete with just a key" scenarios for dynamic
+            `SearchDocument`.  We're using extension methods in .NET but other
+            languages might need to figure out a different approach.
+
+            - `public static class SearchExtensions`
+                - `static void DeleteDocuments(this SearchIndexingBufferedSender<SearchDocument>, string, IEnumerable<string>, CancellationToken)`
```

Comment:
This is an extension method because there is already the `IEnumerable<T> DeleteDocuments` on the buffered sender, right?

Reply:
It's because you can't specialize the `T` to `SearchDocuments` in the generic class.